### ISSUE TITLE
(Willing to PR) "RSS" memory becomes >1GB while other parts are small - is there any utilities to debug this?

### ISSUE THREAD

**Issue by fzyzcjy** [flutter/flutter#128099]
When scrolling a ListView in my app, I see memory going up to 1.3GB. Flutter already has good doc about memory view and profiling in https://docs.flutter.dev/tools/devtools/memory. However, it only talks about how to deal with Dart memory leaks etc, and does not talk about when RSS is huge while Dart memory is small.

I am happy to make PRs (for Flutter tools, Dart tools, etc) if there is a way to examine the RSS memory problem!

![image](https://github.com/flutter/flutter/assets/5236035/52896224-ba80-4420-b631-0201cfaaac30)

**Comment by huycozy** [comment]
In order to address the issue properly, can you please provide a minimal sample code? And please provide the output of `flutter doctor -v` as well. Thanks!

**Comment by fzyzcjy** [comment]
One *possible* reproducible sample (note: this is just *one possibility*, not *all possibilities*, can be seen in #128105. I copy it here:

<details>

```dart
import 'dart:typed_data';
import 'dart:ui';

import 'package:flutter/material.dart' hide Image;

void main() => runApp(const MyApp());

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final images = <Image>[];

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(),
        body: TextButton(
          onPressed: () {
            for (var i = 0; i < 10; ++i) {
              const width = 2000;
              const height = 2000;
              // ref: createTestImage
              decodeImageFromPixels(
                // NOTE this will temporarily create some big array, but will be GCed
                Uint8List.fromList(List<int>.filled(width * height * 4, 0)),
                width,
                height,
                PixelFormat.rgba8888,
                (Image image) {
                  images.add(image);
                  print('leak one image: $image');
                },
              );
            }
          },
          child: Text('leak 10 ui.Image'),
        ),
      ),
    );
  }
}
```

</details>

flutter doctor -v yields latest flutter stable.

**Comment by polina-c** [comment]
Do you use images or pictures in your list?

**Comment by fzyzcjy** [comment]
@polina-c Yes. Indeed I currently suspect it is because I created a bunch of ui.Image without dispose(), and will work on fixing that. On the other hand, this is currently obtained by guessing and trials, so I do hope there are utilities that automatically discover this.

**Comment by huycozy** [comment]
@fzyzcjy Thanks for the sample code, I can see RSS is going up to 0.8-1.2GB with the list of 10 images. 

Adding memory data in profile mode: [dart_devtools_2023-06-05_11_33_34.635.csv.zip](https://github.com/flutter/flutter/files/11648528/dart_devtools_2023-06-05_11_33_34.635.csv.zip)

I'm not sure if this is a documentation issue or it hasn't been implemented on devtools yet (to identify this as a proposal). Labeling the issue for further insights.

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.10.3, on macOS 13.0.1 22A400 darwin-x64, locale en-VN)
    • Flutter version 3.10.3 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f92f44110e (3 days ago), 2023-06-01 18:17:33 -0500
    • Engine revision 2a3401c9bb
    • Dart version 3.0.3
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.78.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.66.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 13.0.1 22A400 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 114.0.5735.90

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

```bash
[!] Flutter (Channel master, 3.11.0-18.0.pre.76, on macOS 13.0.1 22A400 darwin-x64, locale en-VN)
    • Flutter version 3.11.0-18.0.pre.76 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 0b7415356e (2 days ago), 2023-06-03 07:16:37 -0400
    • Engine revision c838a1b059
    • Dart version 3.1.0 (build 3.1.0-166.0.dev)
    • DevTools version 2.24.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.78.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.66.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 13.0.1 22A400 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 114.0.5735.90

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

**Comment by a-siva** [comment]
Maybe this issue needs to be transferred to the flutter/devtools repo.

**Comment by SoftWyer** [comment]
I see similar issues with graphics memory usage when scrolling through a list of items with image thumbnails. 

The images are around 20K in size, however the overall memory increases from an initial 800MB to over 4GB before being killed by the OS on a Pixel 7

Some observations:
* it is the same in debug or release builds
* it affects only some Android devices. A Pixel 4a is OK, total memory usage is steady around 840MB. On a Pixel 7 and a Samsung A53 5G the app will crash having scrolled past ~100 items in the debug build
* the problem exists on the beta channel
* the problem appears to have been resolved in the main channel

It would be great if the fix could be identified and cherry-picked to the stable channel. I'm uncomfortable running off main, but since the app is crashing anyway 🤷 

<details>
  <summary>Flutter doctor v3.10</summary>
  
[√] Flutter (Channel stable, 3.10.5, on Microsoft Windows [Version 10.0.22621.1848], locale en-GB)
    • Flutter version 3.10.5 on channel stable at C:\development\dart\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 796c8ef792 (3 days ago), 2023-06-13 15:51:02 -0700
    • Engine revision 45f6e00911
    • Dart version 3.0.5
    • DevTools version 2.23.1

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at C:\Users\Curly\AppData\Local\Android\sdk
    • Platform android-33, build-tools 33.0.2
    • Java binary at: D:\Program Files\Android\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-9586694)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2022 17.5.5)
    • Visual Studio at d:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.5.33627.172
    • Windows 10 SDK version 10.0.22621.0

[√] Android Studio (version 2022.2)
    • Android Studio at D:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-9586694)

[√] Connected device (4 available)
    • Pixel 7 (mobile)  • 29151FDH2008YH • android-arm64  • Android 14 (API 34)
    • Windows (desktop) • windows        • windows-x64    • Microsoft Windows [Version 10.0.22621.1848]
    • Chrome (web)      • chrome         • web-javascript • Google Chrome 114.0.5735.134
    • Edge (web)        • edge           • web-javascript • Microsoft Edge 114.0.1823.43

[√] Network resources
    • All expected network resources are available.

• No issues found!

</details>

<details>

<summary>Flutter doctor beta channel</summary>

[√] Flutter (Channel beta, 3.12.0, on Microsoft Windows [Version 10.0.22621.1848], locale en-GB)
    • Flutter version 3.12.0 on channel beta at C:\development\dart\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 8fcb74dbc1 (8 days ago), 2023-06-08 13:37:31 -0600
    • Engine revision b84ad1ad30
    • Dart version 3.1.0 (build 3.1.0-163.1.beta)
    • DevTools version 2.24.0

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at C:\Users\Curly\AppData\Local\Android\sdk
    • Platform android-33, build-tools 33.0.2
    • Java binary at: D:\Program Files\Android\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-9586694)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.5.5)
    • Visual Studio at d:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.5.33627.172
    • Windows 10 SDK version 10.0.22621.0

[√] Android Studio (version 2022.2)
    • Android Studio at D:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-9586694)

[√] Connected device (4 available)
    • Pixel 7 (mobile)  • 29151FDH2008YH • android-arm64  • Android 14 (API 34)
    • Windows (desktop) • windows        • windows-x64    • Microsoft Windows [Version 10.0.22621.1848]
    • Chrome (web)      • chrome         • web-javascript • Google Chrome 114.0.5735.134
    • Edge (web)        • edge           • web-javascript • Microsoft Edge 114.0.1823.43

[√] Network resources
    • All expected network resources are available.

• No issues found!
</details>

<details>
  <summary>Flutter doctor main channel</summary>

[√] Flutter (Channel main, 3.12.0-4.0.pre.99, on Microsoft Windows [Version 10.0.22621.1848], locale en-GB)
    • Flutter version 3.12.0-4.0.pre.99 on channel main at C:\development\dart\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fc8856eb80 (40 minutes ago), 2023-06-16 09:53:05 -0400
    • Engine revision fb5fed432e
    • Dart version 3.1.0 (build 3.1.0-213.0.dev)
    • DevTools version 2.24.0

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at C:\Users\Curly\AppData\Local\Android\sdk
    • Platform android-33, build-tools 33.0.2
    • Java binary at: D:\Program Files\Android\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-9586694)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.5.5)
    • Visual Studio at d:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.5.33627.172
    • Windows 10 SDK version 10.0.22621.0

[√] Android Studio (version 2022.2)
    • Android Studio at D:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-9586694)

[√] Connected device (4 available)
    • Pixel 7 (mobile)  • 29151FDH2008YH • android-arm64  • Android 14 (API 34)
    • Windows (desktop) • windows        • windows-x64    • Microsoft Windows [Version 10.0.22621.1848]
    • Chrome (web)      • chrome         • web-javascript • Google Chrome 114.0.5735.134
    • Edge (web)        • edge           • web-javascript • Microsoft Edge 114.0.1823.43

[√] Network resources
    • All expected network resources are available.

• No issues found!

</details>

**Comment by SoftWyer** [comment]
This is still an issue with the **new** stable channel using Flutter v3.13.2. The latest build from the beta channel still exhibits this issue (namely unconstrained graphics memory usage when fast scrolling through a list of cards with thumbnail images on Android).

I've spent some time checking out various builds directly from the main branch and have found that prior to commit [84068857ff707a156e0b70dee8ade2aac792af55](https://github.com/flutter/flutter/commit/84068857ff707a156e0b70dee8ade2aac792af55) the graphics usage for our app on a Pixel 7 is consistent at around 600MB.  With this commit, graphics usage is unconstrained and crashes the app and various other apps as Android OS scavenges memory.

The commit log for this shows:

```
commit 84068857ff707a156e0b70dee8ade2aac792af55
Author: engine-flutter-autoroll <engine-flutter-autoroll@skia.org>
Date:   Fri Jun 30 16:59:13 2023 -0400

    Roll Flutter Engine from 54b573e9c4e5 to e6b8292705a8 (4 revisions) (#129852)

    https://github.com/flutter/engine/compare/54b573e9c4e5...e6b8292705a8

    2023-06-30 jonahwilliams@google.com [Impeller] cache MSAA texture in swapchain for devices that do not support memoryless. (flutter/engine#43349)
    2023-06-30 jonahwilliams@google.com [Impeller] Fix validation error in position color shader. (flutter/engine#43356)
    2023-06-30 109111084+yaakovschectman@users.noreply.github.com Fix inverted boolean in Windows document selection changed a11y event (flutter/engine#43281)
    2023-06-30 bdero@google.com Revert "Reland "add non-rendering operation culling to DisplayListBuilder" (#41463)" (flutter/engine#43358)

```

On main, the problem is fixed with this commit [28dfd836e2390e44e3803573720a1081d724b7b3](https://github.com/flutter/flutter/commit/28dfd836e2390e44e3803573720a1081d724b7b3). The commit log is:

```
commit 28dfd836e2390e44e3803573720a1081d724b7b3
Author: engine-flutter-autoroll <engine-flutter-autoroll@skia.org>
Date:   Wed Jul 19 20:29:23 2023 -0400

    Roll Flutter Engine from 7671e2f2a9fc to eff70f7287f9 (3 revisions) (#130953)

    https://github.com/flutter/engine/compare/7671e2f2a9fc...eff70f7287f9

    2023-07-19 skia-flutter-autoroll@skia.org Roll Dart SDK from 936824d49aa7 to 677bbf64d4d7 (1 revision) (flutter/engine#43833)
    2023-07-19 flar@google.com Reland "Add non-rendering operation culling to DisplayListBuilder" (#41463) (flutter/engine#43831)
    2023-07-19 skia-flutter-autoroll@skia.org Roll Skia from bf75ae2f6eec to 8413c82dea43 (1 revision) (flutter/engine#43832)

```

I'm not quite sure what causes this issue, but it's interesting to see that reverting and then relanding the `DisplayListBuilder` change fixes the issue. The beta branch does not have the relanded `DisplayListBuilder` which might be why that branch is still 'broken'.

It would be nice if someone from the Flutter team could triage this further to see if there's a possible hot fix to stable for this.

**Comment by dnfield** [comment]
You should try to use heapprofd on Android to see what native allocations are happening during this time frame.

**Comment by hnvcam** [comment]
> TL;DR: I'm not sure if this is related to my issue, but while observing the memory chart, I noticed the RSS memory kept increasing to several gigabytes. I'm posting here in case anyone has encountered the same situation and might have a solution.

I was developing a game with Flame and its related libraries (flame_behaviors, flame_riverpod, and flame_audio):
- Whatever I tried, the RSS memory kept increasing, from 2xx MB to 2GB.
- I've tried to reduce everything to minimum: 1 SpriteComponent, 1 overlay Widget for main menu and playing background sound (mp3 file).

No luck, RSS still kept increasing.
- Then I started with profile running on my iPad, played some levels, and I observed that some sound files could not be played because of 'PlatformException(DarwinAudioError....'
- I switched to use https://pub.dev/packages/just_audio, and the above issue has gone.

AND guess what? The RSS memory issue also is gone!!! No more increasing!
I suspects flame_audio causes the issues.

**Comment by WangXiaoRi** [comment]
have encountered the same problem, and various attempts have not been solved. Even an empty secondary page will have an increase in rss memory.
